src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -364,34 +373,262 @@ public void WriteValue(sbyte value)
+        /// according to the specified encoder. It allocates memory for the escaped string and handles
+        /// buffer management using either stack allocation or array pooling. Once the string is escaped,
+        /// it is written into the buffer. If the encoding operation fails, an exception is thrown. 
+        /// </remarks>
+        private void WriteEscapedStringChunk(ReadOnlySpan<char> sourceChunk, int firstIndexToEscape, bool isFinalBlock, out int preChunkNotProcessed)
+        {
+            char[] valueArray = null;
+            const int StackallocCharThreshold = 128;
+
+            int maxSize = firstIndexToEscape + 6 * (sourceChunk.Length - firstIndexToEscape);
+
+            Span<char> destination = maxSize <= StackallocCharThreshold
+                ? stackalloc char[maxSize]
+                : valueArray = ArrayPool<char>.Shared.Rent(maxSize);
+
+            preChunkNotProcessed = 0;
+
+            OperationStatus status = this.encoder.Encode(
+                   sourceChunk,
+                   destination,
+                   out int charsConsumed,
+                   out int charsWritten,
+                   isFinalBlock
+               );
+
+            Debug.Assert(status == OperationStatus.Done || status == OperationStatus.NeedMoreData);
+
+            preChunkNotProcessed = sourceChunk.Length - charsConsumed;
+            WriteStringChunk(destination.Slice(0, charsWritten), isFinalBlock);
+
+            if (valueArray != null)
+            {
+                ArrayPool<char>.Shared.Return(valueArray);
+            }
+        }
+
+        /// <summary>
+        /// Writes a chunk of the string into the buffer, converting it to UTF-8 encoding.
+        /// </summary>
+        /// <param name="chunk">The chunk of the string to be written into the buffer.</param>
+        /// <param name="isFinalBlock">Indicates whether the chunk is the final block of the string.</param>
+        /// <remarks>
+        /// This method converts the provided chunk of the string from UTF-16 encoding to UTF-8 encoding
+        /// and writes it into the buffer. It ensures that the buffer has sufficient space for the encoded
+        /// string and notifies the buffer writer of the write operation. This method assumes that the UTF-8
+        /// encoding does not exceed the maximum byte count specified by the UTF-8 encoding for the given
+        /// chunk of the string.
+        /// </remarks>
+        private void WriteStringChunk(ReadOnlySpan<char> chunk, bool isFinalBlock)
+        {
+            int maxUtf8Length = Encoding.UTF8.GetMaxByteCount(chunk.Length);
+
+            Span<byte> output = bufferWriter.GetSpan(maxUtf8Length);
+            OperationStatus status = Utf8.FromUtf16(chunk, output, out int charsRead, out int charsWritten, isFinalBlock);
+            Debug.Assert(status == OperationStatus.Done || status == OperationStatus.NeedMoreData);
+            Debug.Assert(charsRead == chunk.Length);

```

Comment:
Is `charsRead` guaranteed to be `chunk.Length`? Is it guaranteed that there can be no partial processing of chunks especially for non-final blocks? We do make this consideration for encoding to base64, not sure why we don't do make the same consideration for string to utf8 encoding.

Reply:
To be clear, I'm not sure if that scenario can occur for utf-16 to utf-8 transcoding, but I would request that you confirm first and if it's not an issue for this scenario, then make a comment to explain that.

Reply:
This might be useful: https://learn.microsoft.com/en-us/dotnet/standard/base-types/character-encoding-introduction

Reply:
I think you should add tests with strings that contain surrogate pairs at chunk boundaries.

Reply:
I think we take care of this when we have characters to escape in the chunk... I'm not sure whether we need to do that here too..

Reply:
Let me add tests with surrogate pairs and see how it behaves.

Reply:
I have done some [experiments](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYAMBYAUBgRj1UwAJUCA6AFQFMAPYSgUSgGMB7OASygHMAzpQDqtAEYBuYmQo0GTAFICOUKfhlU6jSgFUo3TnFpriBAJwAKAESAeDcAg+1coAZWv2AALAJQncANwCGYKQcAA7A3CoCpAC8pFC0AO6kSioAyrRg3P4ANtwAXhkA8mERUAJ4AN54pDWshhkxyf4BqWyZYXVcGbpl/gBmtABKtNn+9LRwKVDMAmz+Ibx8eAC+ahSWU+mZOfndW1m5Bdb2VjDBJZGe3nh4APS3AUHKEGBstI22DmqPpM+v77ErIBADcA8ztAwAbO0DAN87EOhQMA+zukIGAbp2EVY1Pcfq56kFYlpgJ0jGA7g9AqRsV1cU0Wm1uB12JSegJ+kMRmMJlMZnMFvxvmSBLRgOF+FFYvEkvjCRl0kLFgILNdcALZSLKABBbLZDgJQb+fi0eV6AxdXX6oQa7KK5XCwTqzXagDC7kC/jYwAy8oA5AAdCAAEQAHJgCJ6zj7/Q6HQAxT2KrEMomNBTNfytdoEhPdB1gWj+d0Wa1yuNkjgQYAhMuNcWkNjOsAAbQAbKQAFS/Uv/ZyuPgeAC66oEqRCeoVfKCOYEEGywEaFKJLEzBY7bwHQ5HnkoqVybwsmDOZk8Z1L5bLR8rvBnnDKEAAthMzzOL6QEpkha4ztwBFHeDkAEJatgAGsQFIPocgFYsggNbkJkaY8KyYLcDFoXczhfbg3ygDdqA4VJgEyfhR1MSwABIrAqCcp2AZYzivSc7zgM4Kjo28Jho59X3dWBSAqdDMOWKxFXWCxoPmCZFTwH4jAEYU81KKtElIMQAE93XrABmVtgjLBCuzcdx+zVQdhygIjcEo6dGh0YA+gDSgozADgb2svoCEbCx4LLTdtxQghD1IaTZPCFQHxrSJWMY7SZz4riPy/H9sn/DggJAsDsggtZzAsMiKINKj2JYhimMKti0M498eJi1wBMVPoOCCCwn24RpMAkUhmoAHg4jCuLa7hEEQTxKmqGphIsWtAk8QKf2CqB624XtvBqUgViAA===) with encoding to try and create a scenario that causes Utf16->utf8 transcoding to result in `NeedMoreData` status, but I think you may be right.

The characters that would cause utf-8 encoding to result in partial processing seem to be already escaped away by the javascriptencoder and the result is written utf8-encoded in full. I have tried different JavaScriptEncoder settings, including allowing specific characters to go unescaped, but the result is the same, the resulting escaped substring is always encoded to utf8 in full.

Maybe you could add a comment to the code to explain why the concerns I raised are not an issue.